[FLINK-30323] Support table statistics in table store

FangYongs opened the pull request:
Support table statistics in table store related #758

JingsongLi reviewed (commented):
Thanks @zjureel for the contribution.

For Flink 1.16, we can consider using `SupportsStatisticReport`.

For the information obtained in the catalog, the `SupportsStatisticReport` returns more accurate information after the Filter Pushdown (such as partition filtering, which is very important), rather than relying on the planner's own inference.

JingsongLi commented on flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/Split.java:
```diff
@@ -23,5 +23,5 @@
 /** An input split for reading. */
 public interface Split extends Serializable {
 
-    long rowCount();
```
Maybe current `rowCount` is OK for now?
`SplitStatistics` is too heavy (column stats are very heavy to compute), we can introduce it in future.
WDYT?

zjureel replied:
Agree, I think the column stats should be optimized instead of reading from file meta directly and we can discuss about it later

JingsongLi commented on flink-table-store-connector/src/main/1.15/org/apache/flink/table/store/connector/source/FlinkTableSource.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.List;
+
+/** A Flink {@link ScanTableSource} for table store. */
+public abstract class FlinkTableSource
```
I was wondering how to avoid copying these codes.
Can we modify `flink.connector.version` to `flink.shim1.version`?
And introduce `flink.shim2.version` and add source through `build-helper` too.

It doesn't look very elegant. In futre, we need to decouple from the flink version. We can only see different versions of flink in the `flink-connector`.

JingsongLi replied:
This one is not resolved?

zjureel replied:
Thanks @JingsongLi , I have fixed it

JingsongLi commented:
I need to reorganize Flink-connector's modules recently, and we can continue to work after that.

JingsongLi commented on flink-table-store-flink/flink-table-store-flink-1.15/src/main/java/org/apache/flink/table/store/connector/source/TableStoreSource.java:
```diff
@@ -0,0 +1,232 @@
+ * FULL scan mode, it will create a {@code HybridSource} of {@link StaticFileStoreSource} and kafka
+ * log source created by {@link LogSourceProvider}.
+ */
+public class TableStoreSource extends FlinkTableSource
```
Can we just introduce a `StatisticTableSource` in flink-common? this can avoid copying for large number of codes.

JingsongLi commented on flink-table-store-flink/flink-table-store-flink-common/src/main/java/org/apache/flink/table/store/connector/source/TableStoreSource.java:
```diff
@@ -209,4 +233,19 @@ public LookupRuntimeProvider getLookupRuntimeProvider(LookupContext context) {
     }
+
+    @Override
+    public TableStats reportStatistics() {
```
Can we document this? This may not be a positive path always. Flink SQL provides `table.optimizer.source.report-statistics-enabled` to close this.

FangYongs commented:
Hi @JingsongLi I have updated this PR, please help to review it again, thanks

JingsongLi commented:
@FangYongs Thanks for the update.

I have thoughts here:
1. We should avoid copying too much code. I think there is a chance to avoid it
2. plan is a heavy work, we can avoid it, we can produce splits in compilation phase, and pass these splits to runtime source. Like in https://github.com/apache/flink-table-store/pull/584

FangYongs commented:
Hi @JingsongLi I have rebase master for this PR, please help to review it when you're free, thanks

JingsongLi reviewed (commented):
We should avoid copying `AbstractFlinkTableFactory`. We can use less codes to fix this.

FangYongs commented:
> We should avoid copying `AbstractFlinkTableFactory`. We can use less codes to fix this.

Then maybe we need to create a factory for DataTableSource, it will create DataTableSource for flink-1.14/flink-1.15 and create StatisticsDataTableSource for the later version of flink. What do you think?

JingsongLi commented:
> > We should avoid copying `AbstractFlinkTableFactory`. We can use less codes to fix this.
> 
> Then maybe we need to create a factory for DataTableSource, it will create DataTableSource for [flink-1](https://issues.apache.org/jira/browse/FLINK-1).14/[flink-1](https://issues.apache.org/jira/browse/FLINK-1).15 and create StatisticsDataTableSource for the later version of flink. What do you think?

We can do like https://github.com/apache/incubator-paimon/pull/1010 , this may not look good, but we definitely need to avoid copying large amounts of code, which can be difficult to maintain.

FangYongs commented:
> > > We should avoid copying `AbstractFlinkTableFactory`. We can use less codes to fix this.
> > 
> > 
> > Then maybe we need to create a factory for DataTableSource, it will create DataTableSource for [flink-1](https://issues.apache.org/jira/browse/FLINK-1).14/[flink-1](https://issues.apache.org/jira/browse/FLINK-1).15 and create StatisticsDataTableSource for the later version of flink. What do you think?
> 
> We can do like #1010 , this may not look good, but we definitely need to avoid copying large amounts of code, which can be difficult to maintain.

Looks good to me, I have rebased master and updated this PR

yuzelin commented on paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/source/statistics/FileStoreTableStatisticsTestBase.java:
```diff
@@ -0,0 +1,181 @@
+    protected FileStoreTable writeData() throws Exception {
+        FileStoreTable table = createStoreTable();
+
+        TableWriteImpl write = table.newWrite(commitUser);
```
Use `TableWriteImpl<?>` to avoid `Raw use` warning in IDE.

FangYongs replied:
Thanks @yuzelin DONE

yuzelin reviewed (commented):
I left a minor improve suggestion, please take a look.

yuzelin reviewed (approved):
+1

JingsongLi reviewed (approved):
+1
